File: frontend/src/engine/hashing.js
import { VECTOR_SIZE } from "./constants";

const TOKEN = /[a-z0-9][a-z0-9_\-./]*/g;

export function tokenize(text) {
  return String(text || "").toLowerCase().match(TOKEN) || [];
}

function fnv1a(str) {
  let hash = 2166136261;
  for (let i = 0; i < str.length; i += 1) {
    hash ^= str.charCodeAt(i);
    hash = Math.imul(hash, 16777619);
  }
  return hash >>> 0;
}

export function embedOne(text, dimensions = VECTOR_SIZE) {
  const tokens = tokenize(text);
  const grams = [...tokens];
  for (let i = 0; i < tokens.length - 1; i += 1) {
    grams.push(`${tokens[i]}_${tokens[i + 1]}`);
  }
  const vec = new Array(dimensions).fill(0);
  if (!grams.length) return vec;
  for (const gram of grams) {
    const number = fnv1a(gram);
    const index = number % dimensions;
    const sign = (number >>> 20) & 1 ? -1 : 1;
    vec[index] += sign;
  }
  const norm = Math.sqrt(vec.reduce((sum, value) => sum + value * value, 0)) || 1;
  return vec.map((value) => value / norm);
}

export function cosine(a, b) {
  let sum = 0;
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i += 1) sum += a[i] * b[i];
  return sum;
}

export function contentHash(text) {
  const normalized = String(text || "")
    .trim()
    .split(/\n/)
    .map((line) => line.replace(/\s+$/g, ""))
    .join("\n");
  return fnv1a(normalized).toString(16).padStart(8, "0") + fnv1a(`:${normalized.length}`).toString(16);
}

File: backend/app/providers/embeddings/hashing.py
"""Embedding provider: hashed n-gram vectors.

This is a deterministic development fallback. Same token always lands in
the same dimension, so overlapping vocabulary produces similar vectors.
It is NOT a neural semantic embedding model and must not be described as one.
"""

from __future__ import annotations

import hashlib
import math
import re

from app.core.constants import VECTOR_SIZE

_TOKEN = re.compile(r"[a-z0-9][a-z0-9_\-./]*")


def tokenize(text: str) -> list[str]:
    return _TOKEN.findall(text.lower())


class HashingEmbeddingProvider:
    name = "dev-hashing"
    dimensions = VECTOR_SIZE
    description = (
        "Deterministic hashed n-gram embeddings for local development. "
        "Not a production semantic model."
    )

    def __init__(self, dimensions: int = VECTOR_SIZE) -> None:
        self.dimensions = dimensions

    def embed_one(self, text: str) -> list[float]:
        tokens = tokenize(text)
        grams = list(tokens)
        grams.extend(f"{a}_{b}" for a, b in zip(tokens, tokens[1:], strict=False))
        vec = [0.0] * self.dimensions
        if not grams:
            return vec
        for gram in grams:
            digest = hashlib.blake2b(gram.encode("utf-8"), digest_size=8).digest()
            number = int.from_bytes(digest, "little")
            index = number % self.dimensions
            sign = 1.0 if (number >> 20) & 1 == 0 else -1.0
            vec[index] += sign
        norm = math.sqrt(sum(value * value for value in vec)) or 1.0
        return [value / norm for value in vec]

    async def embed(self, texts: list[str]) -> list[list[float]]:
        return [self.embed_one(text) for text in texts]
